README.md
[![Build Status](https://travis-ci.org/nth10sd/search-keys.svg?branch=master)](https://travis-ci.org/nth10sd/search-keys/)

# Search Keys
Now ported to WebExtensions, originally a [legacy add-on](https://www.squarefree.com/extensions/search-keys/):

```
For example, pressing 1 takes you to the first result. Hold Alt (Windows/Mac) or Ctrl (Linux) to open results in new tabs, or Shift to open results in new windows.

This extension can save you time because your hands are usually on the keyboard after you enter a search.
```

Search Keys supports Google, Yahoo HK/TW/US, Bing, DuckDuckGo and Wikipedia.

# Requirements
Tested in Firefox 55 and up.

# Known Bugs
* When you use Search Keys to follow a link, the referrer is not sent.
* If a search result's title text has a direction different from the page's direction (e.g. a Hebrew result in an English page), the green box indicating the shortcut is misplaced.
* On the Hebrew-language version of Google, the hint should be placed to the left but is placed to the right.
* Doesn't work when the search-results page is in a frame.

# Screenshots
* Google:
![Screenshot of Google Search with search numbers][google-search]

* Yahoo (US)
![Screenshot of Yahoo Search (US) with search numbers][yahoo-us-search]

* Bing
![Screenshot of Bing Search with search numbers][bing-search]

[google-search]: https://github.com/nth10sd/search-keys/blob/master/screenshots/google.png "Screenshot of Google Search with search numbers"
[yahoo-us-search]: https://github.com/nth10sd/search-keys/blob/master/screenshots/yahoo-us.png "Screenshot of Yahoo Search (US) with search numbers"
[bing-search]: https://github.com/nth10sd/search-keys/blob/master/screenshots/bing.png "Screenshot of Bing Search with search numbers"


CHANGELOG.md
# Changelog

* 2.1.0 (2017-11-02)
  * Unforked, returning to the "Search Keys" name
  * Renamed back all references to "Add Search Number"
  * Added a "package-lock.json" file for npm 5 usage

* 2.0.0 (2017-09-13)
  * Initially forked as "Add Search Number"
  * Finally moved off old legacy interfaces, e.g. nsIIOService
  * Make search result link detection use querySelector
  * We no longer focus on elements when they are selected as the focus code was not working well
  * Reinstated/Added support for most major default search engines
    * Yahoo HK/TW/US
    * Bing
    * DuckDuckGo
    * Wikipedia

* 1.9.0 beta (2017-09-08) - unreleased
  * Forked from [Search Keys](https://www.squarefree.com/extensions/search-keys/) version 1.0.1
  * Name changed to "Add Search Number"
  * Ported to WebExtensions
  * Beta for now, until it is stable enough to hit 2.0.0

The following is adapted from an old [changelog](https://www.squarefree.com/extensions/search-keys/). All links updated to use https or point to an archive.org version where unavailable.

* Search Keys 1.0.1 (2010-01-05)
  * Fix JavaScript strict warnings

* Search Keys 1.0 (2009-11-05)
  * Updated to work with Google's next/previous page   links again
  * Updated to work on Yahoo again
  * Added Bing
  * Removed many engines that broke over the years.
  * Unjarred chrome
  * Switched from `contents.rdf` to `chrome.manifest`
    * Required for Firefox 3.6

* Search Keys 0.8.1 (2006-11-28)
  * Updated to work on Google again.
  * Dropped support for Google Groups.
    * The site recently changed to use less semantic HTML
      * Made it harder for Search Keys to identify the main links
    * Also, there's a beta, which uses different HTML
      * May presumably replace Google Groups at some point.

* Search Keys 0.8 (2006-10-28)
  * Updated to make the previous/next shortcuts (`,` and `.`) work on Google again
    * Google made a clever change
      * The navigation section no longer uses a bunch of tiny images
      * Instead splits [this image](https://www.google.com/images/nav_logo.png) using CSS
  * Updated to work on Google News again
  * Updated to work on Yahoo Local again
  * Removed support for Google Local, which has been merged in Google Maps
    * These shortcuts don't make much sense on Google Maps
  * Made the hints show up on trunk again
    * Using the technique in [this article](https://developer.mozilla.org/en-US/Add-ons/Code_snippets/On_page_load)
  * Fixed it to create DOM nodes for the hints in the correct document
    * So it won't break when [Firefox trunk starts throwing WRONG_DOCUMENT_ERR](https://bugzilla.mozilla.org/show_bug.cgi?id=47903) in compliance with DOM specs

* Search Keys 0.7.3 (2006-01-13)
  * Fixed a potential minor leak by adding `var link;` to the goToResult function

* Search Keys 0.7.2 (2005-11-29)
  * Made it work on pages resulting from address bar searches
    * You'll only notice this fix if you have [changed keyword.URL](https://www.squarefree.com/2004/09/09/googles-browse-by-name-in-firefox/) to something other than Google's "I'm Feeling Lucky"
    * The fix is a workaround for a bug in Firefox, [bug 264830](https://bugzilla.mozilla.org/show_bug.cgi?id=264830)

* Search Keys 0.7.1 (2005-11-28)
  * Fixed it to work on Google again
    * Broken by a recent change by Google
  * Fixed it to work with del.icio.us again
    * Broken by a not-so-recent change by del.icio.us

* Search Keys 0.7
  * Fixed a JavaScript Strict warning: `function getActiveEngine does not always return a value`
  * Made `.` and `,` navigate between pages of search results on Google
  * Marked the extension as working in Firefox 1.1α1

* Search Keys 0.6 (2004-11-13)
  * Add support for [Google Local](https://local.google.com/)
  * Add support for [Yahoo!](https://search.yahoo.com/), [Yahoo! Image Search](https://images.search.yahoo.com/), [Yahoo! News search](https://news.yahoo.com/), [Yahoo! Local](https://local.yahoo.com/), and [Yahoo! Personals](https://personals.yahoo.com/)
    * [Code provided by Marc Abramowitz of Yahoo!](https://web.archive.org/web/20081211234854/http://www.ysearchblog.com/archives/000044.html)

* Search Keys 0.5 (2004-10-29)
  * Make number keys on numeric keypad work
    * They only work when Numlock is on and they don't work with Shift

* Search Keys 0.4.2 (2004-10-16)
  * Support custom Google web searches in addition to normal Google searches
  * Support Google Images, with no boxes showing shortcuts

* Search Keys 0.4.1 (2004-10-16)
  * Also avoid conflict with tab-switching shortcut on Linux

* Search Keys 0.4 (2004-10-16)
  * On Windows and Mac, use Alt+number instead of Ctrl+number for opening in a new tab
    * To avoid conflict with tab-switching shortcuts
  * Skip preview links added by the Search Preview and McSearchPreview extensions
  * Remove Bugzilla support
  * Update maxVersion so Firefox 1.0 knows this extension works with it

* Search Keys 0.3 (2004-10-15)
  * Support international versions of Google, Google News, and Google Groups
  * Make boxes containing numbers slightly shorter
  * Add an extension icon

* Search Keys 0.2 (2004-10-14)
  * Next to each search result, show its number

* Search Keys 0.1 (2004-10-14)
  * Initial release


background.js
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

function getMessage(message) {
  if (message.selection == "current") {
    // Navigate to the URL in the current tab
    browser.tabs.update({
      active: true,
      url: message.url
    })
  } else if (message.selection == "tab") {
    // Navigate to the URL in a new tab
    browser.tabs.create({
      active: true,
      // openerTabId: XX,  # Wait for Firefox 57, see bug 1238314
      url: message.url
    })
  } else if (message.selection == "window") {
    // Navigate to the URL in a new window
    browser.windows.create({
      url: message.url
    })
  }
}


// function getActiveTab() {
//   return browser.tabs.query({
//     active: true,
//     currentWindow: true
//   });
// }


// function gotPlatformInfo(info) {
//   getActiveTab().then((tabs) => {
//     browser.tabs.sendMessage(tabs[0].id, {
//       platform: info.os
//     })
//   })
// }


browser.runtime.onMessage.addListener(getMessage);
// browser.runtime.getPlatformInfo().then(gotPlatformInfo);


searchkeys.js
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

var searchkeysEngines = [
  // Each search engine has two boolean functions:
  //
  // * test(url).  Returns true if this URL is a search results page for this search engine.
  // * testLink(linkNode).  Returns true if this link represents a search result.
  //
  // Two optional functions return link nodes:
  //
  // * prev(doc).  Returns an object with the link to the previous page of search results and the node to be clicked,
  //               or null if none.
  //
  // * next(doc).  Returns an object with the link to the next page of search results and the node to be clicked,
  //               or null if none.
  {
    // results in the "did you mean?" section have no className. other results have classname of l (lowercase L).

    name: "Google (web search)",
    test: function (url) {
      return url.host.indexOf("google") !== -1 && url.pathname.indexOf("/search") !== -1;
    },
    testLink: function (linkNode) {
      return (linkNode.className === "l" || linkNode.className === "") && // empty for did-you-mean results (desired)
        linkNode.parentNode.tagName.toLowerCase() === "h3" && // h4 for local maps results (not desired)
        linkNode.parentNode.className === "r";
    },
    prev: function (document) {
      var pNode = document.querySelector("a[id='pnprev']");
      if (!pNode)
        return null;
      return {
        "href": pNode.href,
        "clickNode": pNode.lastElementChild
      }
    },
    next: function (document) {
      var nNode = document.querySelector("a[id='pnnext']");
      if (!nNode)
        return null;
      return {
        "href": nNode.href,
        "clickNode": nNode.lastElementChild
      }
    },
  },
  {
    // Works with Yahoo HK/TW/US
    name: "Yahoo (web search)",
    test: function (url) {
      return url.host.indexOf("yahoo") !== -1 && url.pathname.indexOf("/search") !== -1;
    },
    testLink: function (linkNode) {
      // Yahoo has a bunch of randomly generated class names in the link nodes itself
      return (linkNode.parentNode.parentNode.className.indexOf("comp") !== -1 &&
          linkNode.parentNode.parentNode.className.toLowerCase().indexOf("title") !== -1) &&
        // Ignore image and video results
        (linkNode.parentNode.parentNode.parentNode.className.indexOf("dd algo") !== -1 ||
          linkNode.parentNode.parentNode.parentNode.parentNode.className.indexOf("dd algo") !== -1) &&
        (linkNode.parentNode.tagName.toLowerCase() === "h3" || linkNode.parentNode.tagName.toLowerCase() === "p") &&
        (linkNode.parentNode.className === "title" || linkNode.parentNode.className === "");
    },
    prev: function (document) {
      var nNode = document.querySelector("a.prev");
      if (!nNode)
        return null;
      return {
        "href": nNode.href,
        "clickNode": nNode
      }
    },
    next: function (document) {
      var nNode = document.querySelector("a.next");
      if (!nNode)
        return null;
      return {
        "href": nNode.href,
        "clickNode": nNode
      }
    },
  },
  {
    name: "Bing (web search)",
    test: function (url) {
      return url.host.indexOf("bing") !== -1 && url.pathname.indexOf("/search") !== -1;
    },
    testLink: function (linkNode) {
      return linkNode.parentNode.tagName.toLowerCase() === "h2" &&
        linkNode.parentNode.parentNode.className === "b_algo";
    },
    prev: function (document) {
      var nNode = document.querySelector("a.sb_pagP");
      if (!nNode)
        return null;
      return {
        "href": nNode.href,
        "clickNode": nNode.parentNode
      }
    },
    next: function (document) {
      var nNode = document.querySelector("a.sb_pagN");
      if (!nNode)
        return null;
      return {
        "href": nNode.href,
        "clickNode": nNode.parentNode
      }
    },
  },
  {
    name: "DuckDuckGo (web search)",
    test: function (url) {
      return url.host.indexOf("duckduckgo") !== -1 && url.search.indexOf("?q=") !== -1;
    },
    testLink: function (linkNode) {
      return linkNode.parentNode.querySelector("a.badge--ad") === null && // Ignore ads
        linkNode.className === "result__a" &&
        linkNode.parentNode.className === "result__title";
    },
    // DuckDuckGo continuously displays search results as you scroll, so support for "prev" and "next" are removed
  },
  {
    name: "Wikipedia search",
    test: function (url) {
      return url.host.indexOf("wikipedia") !== -1 && url.search.toLowerCase().indexOf("special:search") !== -1;
    },
    testLink: function (linkNode) {
      return linkNode.parentNode.className === "mw-search-result-heading";
    },
    prev: function (document) {
      var nNode = document.querySelector("a.mw-prevlink");
      if (!nNode)
        return null;
      return {
        "href": nNode.href,
        "clickNode": nNode
      }
    },
    next: function (document) {
      var nNode = document.querySelector("a.mw-nextlink");
      if (!nNode)
        return null;
      return {
        "href": nNode.href,
        "clickNode": nNode
      }
    },
  },
];

// With Shift+1, keydown gets "1" (keyCode-48) while keypress gets "!" (charCode) with an American keyboard.
// But we have to use keypress to suppress FAYT.
// Crazy solution: assume keypress happens only immediately after keydown.  Have onkeydown set
// a global variable, suppressKeypress, to tell onkeypress whether to stopPropagation and stuff.
//
// https://bugzilla.mozilla.org/show_bug.cgi?id=167145

var suppressKeypress = false;

window.addEventListener("keydown", searchkeysKeydown, true); // Use capturing to beat FAYT.
window.addEventListener("keypress", searchkeysKeypress, true); // Use capturing to beat FAYT.


function init() {
  var engine = getActiveEngine(document);
  if (!engine)
    return null;

  var currentResultNumber = 0;

  var i, link;
  for (i = 0;
    (link = document.links[i]); ++i) { // Warning: loop is very similar to a loop in another function
    if (engine.testLink(link)) {
      ++currentResultNumber;
      addHint(link, currentResultNumber);
      if (currentResultNumber == 10)
        break;
    }
  }

  if (engine.next) {
    var next = engine.next(document);
    if (next)
      next.clickNode.appendChild(document.createTextNode(" (.)"));
  }

  if (engine.prev) {
    var prev = engine.prev(document);
    if (prev)
      prev.clickNode.appendChild(document.createTextNode(" (,)"));
  }
}

init();


function ensureURL(urllink) {
  var url;
  url = new URL(urllink);

  if (!url) return null;

  if (url.protocol.indexOf("http") !== -1 || url.protocol.indexOf("https") !== -1) {
    return url;
  } else {
    return null;
  }
}


function keycodeToTarget(keyCode) {
  // Number keys above letter keys
  if (48 < keyCode && keyCode <= 57) return keyCode - 48;
  if (keyCode == 48) return 10; // zero

  // Numeric keypad, with numlock on
  if (96 < keyCode && keyCode <= 105) return keyCode - 96;
  if (keyCode == 96) return 10; // zero

  if (keyCode == 188) // , or <
    return "prev";

  if (keyCode == 190) // . or >
    return "next";

  // Any other key
  return null;
}


function searchkeysKeydown(event) {
  suppressKeypress = false;

  // A number from 1 to 10, "next", "prev", or null.
  var target = keycodeToTarget(event.keyCode);

  // Only trigger for digits, comma, and period.
  if (target == null)
    return;

  // Don't trigger if a form element has focus.
  var elt = document.activeElement;
  if (elt) {
    var ln = new String(elt.localName).toLowerCase();
    if (ln == "input" || ln == "textarea" || ln == "select" || ln == "isindex")
      return;
  }

  // Only trigger for modifiers we should trigger for. (And calculate |where| for later.)
  var where = whereToOpen(event);
  if (where == "ignore")
    return; // in particular, don't suppress the keypress

  var activeEngine = getActiveEngine(document);
  if (!activeEngine)
    return;

  // Go there.
  goToResult(activeEngine, target, where);

  // Suppress the keypress associated with this keydown, to stop FAYT.
  suppressKeypress = true;
}


function searchkeysKeypress(event) {
  if (suppressKeypress) {
    // We handled the event, discourage others from handling it too.
    event.preventDefault(); // seems like a good idea
    event.stopPropagation(); // stop FAYT. would like to just stop FAYT from *starting*, but oh well.
  }
}


// Based on utilityOverlay.js from Firefox.
function whereToOpen(e) {
  // This isn't working yet. This will likely improve cross-browser support.
  //
  // var currentPlatform = "";
  // browser.runtime.onMessage.addListener(request => {
  //   currentPlatform = request.platform;
  // });
  // var win = (currentPlatform.indexOf("win") !== -1);
  // var mac = (currentPlatform.indexOf("mac") !== -1);

  // Platform differences suck.
  // Really want to say "#ifndef XP_UNIX" from 256635, but I don't know how to do that in JS.
  var win = (navigator.platform.indexOf("Win") != -1);
  var mac = (navigator.platform.indexOf("Mac") != -1);
  var altForTabs = win || mac;

  var shift = e.shiftKey;
  var ctrl = e.ctrlKey;
  // var meta =  e.metaKey;
  var alt = e.altKey;

  if (altForTabs ? (alt && !ctrl) : (ctrl && !alt)) { /* Mac normally uses cmd (meta), btw */
    if (shift)
      return "tabshifted";
    else
      return "tab";
  } else if (!altForTabs ? (alt && !ctrl) : (ctrl && !alt)) {
    return "ignore";
  } else if (alt && ctrl) {
    // Bug 69954 prevents this from being reached on Windows.
    return null;
  } else if (shift) {
    return "window";
  } else {
    return "current";
  }
}


function addHint(linkNode, resultNumber) {
  var doc = linkNode.ownerDocument;

  var hint = doc.createElementNS("http://www.w3.org/1999/xhtml", "span");
  hint.className = "searchkeys-hint" // for the benefit of user style sheets, test styles
  hint.style.color = "green";
  hint.style.background = "white";
  hint.style.border = "1px solid";
  hint.style.padding = "1px 2px 1px 2px";
  hint.style.marginLeft = ".5em"; // inline elements can have margins? cool!
  // hint.style.marginRight = ".5em"; // just in case we're in RTL-land?

  hint.appendChild(doc.createTextNode(resultNumber == 10 ? 0 : resultNumber));

  linkNode.parentNode.insertBefore(hint, linkNode.nextSibling);
}


function getActiveEngine(doc) {
  if (!doc.location) // when would this happen? it did...
    return null;

  var url = ensureURL(doc.location.href);

  if (!url)
    return null;

  var i, engine;
  for (i = 0;
    (engine = searchkeysEngines[i]); ++i)
    if (engine.test(url))
      return engine;

  return null;
}


function goToResult(engine, resultNumber, where) {
  var link;

  if (resultNumber == "next" && engine.next) {
    link = engine.next(document);
  } else if (resultNumber == "prev" && engine.prev) {
    link = engine.prev(document);
  } else {
    link = findResultNumbered(engine, resultNumber);
  }

  if (link) {
    var urlout = ensureURL(link.href);

    if (!urlout)
      return;

    // Follow the link, but note that no referrer is sent.
    if (where == "current") {
      browser.runtime.sendMessage({
        "selection": where,
        "url": urlout.href
      });
    } else if (where == "tab") {
      browser.runtime.sendMessage({
        "selection": where,
        "url": urlout.href
      });
    } else if (where == "window") {
      browser.runtime.sendMessage({
        "selection": where,
        "url": urlout.href
      });
    } else return;
  } else return; // Might be an absent "prev" or "next"
}


function findResultNumbered(engine, resultNumber) {
  var links = document.links;

  var currentResultNumber = 0;

  var i, link;
  for (i = 0;
    (link = links[i]); ++i) { // Warning: loop is very similar to a loop in another function
    if (engine.testLink(link)) {
      ++currentResultNumber;
      if (currentResultNumber == resultNumber) {
        return link;
      }
    }
  }

  return null;
}


addsearchnumber.js
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

var searchnumbersEngines = [
  // Each search engine has two boolean functions:
  //
  // * test(url).  Returns true if this URL is a search results page for this search engine.
  // * testLink(linkNode).  Returns true if this link represents a search result.
  //
  // Two optional functions return link nodes:
  //
  // * prev(doc).  Returns an object with the link to the previous page of search results and the node to be clicked,
  //               or null if none.
  //
  // * next(doc).  Returns an object with the link to the next page of search results and the node to be clicked,
  //               or null if none.
  {
    // results in the "did you mean?" section have no className. other results have classname of l (lowercase L).

    name: "Google (web search)",
    test: function (url) {
      return url.host.indexOf("google") !== -1 && url.pathname.indexOf("/search") !== -1;
    },
    testLink: function (linkNode) {
      return (linkNode.className === "l" || linkNode.className === "") && // empty for did-you-mean results (desired)
        linkNode.parentNode.tagName.toLowerCase() === "h3" && // h4 for local maps results (not desired)
        linkNode.parentNode.className === "r";
    },
    prev: function (document) {
      var pNode = document.querySelector("a[id='pnprev']");
      if (!pNode)
        return null;
      return {
        "href": pNode.href,
        "clickNode": pNode.lastElementChild
      }
    },
    next: function (document) {
      var nNode = document.querySelector("a[id='pnnext']");
      if (!nNode)
        return null;
      return {
        "href": nNode.href,
        "clickNode": nNode.lastElementChild
      }
    },
  },
  {
    // Works with Yahoo HK/TW/US
    name: "Yahoo (web search)",
    test: function (url) {
      return url.host.indexOf("yahoo") !== -1 && url.pathname.indexOf("/search") !== -1;
    },
    testLink: function (linkNode) {
      // Yahoo has a bunch of randomly generated class names in the link nodes itself
      return (linkNode.parentNode.parentNode.className.indexOf("comp") !== -1 &&
          linkNode.parentNode.parentNode.className.toLowerCase().indexOf("title") !== -1) &&
        // Ignore image and video results
        (linkNode.parentNode.parentNode.parentNode.className.indexOf("dd algo") !== -1 ||
          linkNode.parentNode.parentNode.parentNode.parentNode.className.indexOf("dd algo") !== -1) &&
        (linkNode.parentNode.tagName.toLowerCase() === "h3" || linkNode.parentNode.tagName.toLowerCase() === "p") &&
        (linkNode.parentNode.className === "title" || linkNode.parentNode.className === "");
    },
    prev: function (document) {
      var nNode = document.querySelector("a.prev");
      if (!nNode)
        return null;
      return {
        "href": nNode.href,
        "clickNode": nNode
      }
    },
    next: function (document) {
      var nNode = document.querySelector("a.next");
      if (!nNode)
        return null;
      return {
        "href": nNode.href,
        "clickNode": nNode
      }
    },
  },
  {
    name: "Bing (web search)",
    test: function (url) {
      return url.host.indexOf("bing") !== -1 && url.pathname.indexOf("/search") !== -1;
    },
    testLink: function (linkNode) {
      return linkNode.parentNode.tagName.toLowerCase() === "h2" &&
        linkNode.parentNode.parentNode.className === "b_algo";
    },
    prev: function (document) {
      var nNode = document.querySelector("a.sb_pagP");
      if (!nNode)
        return null;
      return {
        "href": nNode.href,
        "clickNode": nNode.parentNode
      }
    },
    next: function (document) {
      var nNode = document.querySelector("a.sb_pagN");
      if (!nNode)
        return null;
      return {
        "href": nNode.href,
        "clickNode": nNode.parentNode
      }
    },
  },
  {
    name: "DuckDuckGo (web search)",
    test: function (url) {
      return url.host.indexOf("duckduckgo") !== -1 && url.search.indexOf("?q=") !== -1;
    },
    testLink: function (linkNode) {
      return linkNode.parentNode.querySelector("a.badge--ad") === null && // Ignore ads
        linkNode.className === "result__a" &&
        linkNode.parentNode.className === "result__title";
    },
    // DuckDuckGo continuously displays search results as you scroll, so support for "prev" and "next" are removed
  },
  {
    name: "Wikipedia search",
    test: function (url) {
      return url.host.indexOf("wikipedia") !== -1 && url.search.toLowerCase().indexOf("special:search") !== -1;
    },
    testLink: function (linkNode) {
      return linkNode.parentNode.className === "mw-search-result-heading";
    },
    prev: function (document) {
      var nNode = document.querySelector("a.mw-prevlink");
      if (!nNode)
        return null;
      return {
        "href": nNode.href,
        "clickNode": nNode
      }
    },
    next: function (document) {
      var nNode = document.querySelector("a.mw-nextlink");
      if (!nNode)
        return null;
      return {
        "href": nNode.href,
        "clickNode": nNode
      }
    },
  },
];

// With Shift+1, keydown gets "1" (keyCode-48) while keypress gets "!" (charCode) with an American keyboard.
// But we have to use keypress to suppress FAYT.
// Crazy solution: assume keypress happens only immediately after keydown.  Have onkeydown set
// a global variable, suppressKeypress, to tell onkeypress whether to stopPropagation and stuff.
//
// https://bugzilla.mozilla.org/show_bug.cgi?id=167145

var suppressKeypress = false;

window.addEventListener("keydown", searchnumbersKeydown, true); // Use capturing to beat FAYT.
window.addEventListener("keypress", searchnumbersKeypress, true); // Use capturing to beat FAYT.


function init() {
  var engine = getActiveEngine(document);
  if (!engine)
    return null;

  var currentResultNumber = 0;

  var i, link;
  for (i = 0;
    (link = document.links[i]); ++i) { // Warning: loop is very similar to a loop in another function
    if (engine.testLink(link)) {
      ++currentResultNumber;
      addHint(link, currentResultNumber);
      if (currentResultNumber == 10)
        break;
    }
  }

  if (engine.next) {
    var next = engine.next(document);
    if (next)
      next.clickNode.appendChild(document.createTextNode(" (.)"));
  }

  if (engine.prev) {
    var prev = engine.prev(document);
    if (prev)
      prev.clickNode.appendChild(document.createTextNode(" (,)"));
  }
}

init();


function ensureURL(urllink) {
  var url;
  url = new URL(urllink);

  if (!url) return null;

  if (url.protocol.indexOf("http") !== -1 || url.protocol.indexOf("https") !== -1) {
    return url;
  } else {
    return null;
  }
}


function keycodeToTarget(keyCode) {
  // Number keys above letter keys
  if (48 < keyCode && keyCode <= 57) return keyCode - 48;
  if (keyCode == 48) return 10; // zero

  // Numeric keypad, with numlock on
  if (96 < keyCode && keyCode <= 105) return keyCode - 96;
  if (keyCode == 96) return 10; // zero

  if (keyCode == 188) // , or <
    return "prev";

  if (keyCode == 190) // . or >
    return "next";

  // Any other key
  return null;
}


function searchnumbersKeydown(event) {
  suppressKeypress = false;

  // A number from 1 to 10, "next", "prev", or null.
  var target = keycodeToTarget(event.keyCode);

  // Only trigger for digits, comma, and period.
  if (target == null)
    return;

  // Don't trigger if a form element has focus.
  var elt = document.activeElement;
  if (elt) {
    var ln = new String(elt.localName).toLowerCase();
    if (ln == "input" || ln == "textarea" || ln == "select" || ln == "isindex")
      return;
  }

  // Only trigger for modifiers we should trigger for. (And calculate |where| for later.)
  var where = whereToOpen(event);
  if (where == "ignore")
    return; // in particular, don't suppress the keypress

  var activeEngine = getActiveEngine(document);
  if (!activeEngine)
    return;

  // Go there.
  goToResult(activeEngine, target, where);

  // Suppress the keypress associated with this keydown, to stop FAYT.
  suppressKeypress = true;
}


function searchnumbersKeypress(event) {
  if (suppressKeypress) {
    // We handled the event, discourage others from handling it too.
    event.preventDefault(); // seems like a good idea
    event.stopPropagation(); // stop FAYT. would like to just stop FAYT from *starting*, but oh well.
  }
}


// Based on utilityOverlay.js from Firefox.
function whereToOpen(e) {
  // This isn't working yet. This will likely improve cross-browser support.
  //
  // var currentPlatform = "";
  // browser.runtime.onMessage.addListener(request => {
  //   currentPlatform = request.platform;
  // });
  // var win = (currentPlatform.indexOf("win") !== -1);
  // var mac = (currentPlatform.indexOf("mac") !== -1);

  // Platform differences suck.
  // Really want to say "#ifndef XP_UNIX" from 256635, but I don't know how to do that in JS.
  var win = (navigator.platform.indexOf("Win") != -1);
  var mac = (navigator.platform.indexOf("Mac") != -1);
  var altForTabs = win || mac;

  var shift = e.shiftKey;
  var ctrl = e.ctrlKey;
  // var meta =  e.metaKey;
  var alt = e.altKey;

  if (altForTabs ? (alt && !ctrl) : (ctrl && !alt)) { /* Mac normally uses cmd (meta), btw */
    if (shift)
      return "tabshifted";
    else
      return "tab";
  } else if (!altForTabs ? (alt && !ctrl) : (ctrl && !alt)) {
    return "ignore";
  } else if (alt && ctrl) {
    // Bug 69954 prevents this from being reached on Windows.
    return null;
  } else if (shift) {
    return "window";
  } else {
    return "current";
  }
}


function addHint(linkNode, resultNumber) {
  var doc = linkNode.ownerDocument;

  var hint = doc.createElementNS("http://www.w3.org/1999/xhtml", "span");
  hint.className = "searchnumbers-hint" // for the benefit of user style sheets, test styles
  hint.style.color = "green";
  hint.style.background = "white";
  hint.style.border = "1px solid";
  hint.style.padding = "1px 2px 1px 2px";
  hint.style.marginLeft = ".5em"; // inline elements can have margins? cool!
  // hint.style.marginRight = ".5em"; // just in case we're in RTL-land?

  hint.appendChild(doc.createTextNode(resultNumber == 10 ? 0 : resultNumber));

  linkNode.parentNode.insertBefore(hint, linkNode.nextSibling);
}


function getActiveEngine(doc) {
  if (!doc.location) // when would this happen? it did...
    return null;

  var url = ensureURL(doc.location.href);

  if (!url)
    return null;

  var i, engine;
  for (i = 0;
    (engine = searchnumbersEngines[i]); ++i)
    if (engine.test(url))
      return engine;

  return null;
}


function goToResult(engine, resultNumber, where) {
  var link;

  if (resultNumber == "next" && engine.next) {
    link = engine.next(document);
  } else if (resultNumber == "prev" && engine.prev) {
    link = engine.prev(document);
  } else {
    link = findResultNumbered(engine, resultNumber);
  }

  if (link) {
    var urlout = ensureURL(link.href);

    if (!urlout)
      return;

    // Follow the link, but note that no referrer is sent.
    if (where == "current") {
      browser.runtime.sendMessage({
        "selection": where,
        "url": urlout.href
      });
    } else if (where == "tab") {
      browser.runtime.sendMessage({
        "selection": where,
        "url": urlout.href
      });
    } else if (where == "window") {
      browser.runtime.sendMessage({
        "selection": where,
        "url": urlout.href
      });
    } else return;
  } else return; // Might be an absent "prev" or "next"
}


function findResultNumbered(engine, resultNumber) {
  var links = document.links;

  var currentResultNumber = 0;

  var i, link;
  for (i = 0;
    (link = links[i]); ++i) { // Warning: loop is very similar to a loop in another function
    if (engine.testLink(link)) {
      ++currentResultNumber;
      if (currentResultNumber == resultNumber) {
        return link;
      }
    }
  }

  return null;
}
